### ISSUE TITLE
Mailpile now available as 1.0rc1

### ISSUE THREAD

**Issue by jeaye** [NixOS/nixpkgs#28111]
## Issue description
The [existing Mailpile package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/mailreaders/mailpile/default.nix) has been untouched for a while, since the Mailpile devs [suggested not to use it yet](https://github.com/NixOS/nixpkgs/pull/23058#issuecomment-283515104). Now, it's [available for Debian](https://www.mailpile.is/download/linux.html) and was announced at [SHA2017](https://www.youtube.com/watch?v=RgAIlqV-OYk) that it's ready for 1.0rc1.

It'd be great to have both the package and [service](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/mailpile.nix) updated to follow the release candidates.

**Comment by volth** [comment]
SHA2017 link says: "This video has been removed by the user"

**Comment by jeaye** [comment]
Updated the link.

**Comment by jeaye** [comment]
Blog post about it is here: https://www.mailpile.is/blog/2017-08-13_SHA2017.html

**Comment by BjarniRunar** [comment]
Note that we're still not looking to get included in distributions - we're running our own repositories and our own packaging robot, to be sure bugfixes get to people in a timely fashion. If there is interest here, would it make sense to teach our bot to build NixOS packages?

**Comment by domenkozar** [comment]
@BjarniRunar that sounds good, how do we go and do that? Note that it requires root Nix install to work.

**Comment by BjarniRunar** [comment]
@domenkozar The way I build the Debian packages is largely based on Docker; if I had a Dockerfile that would build Mailpile packages for NixOS then that would be relatively easy to add to my build-bot.

**Comment by MostAwesomeDude** [comment]
`FROM nixos/nix` will get you started. Here's a Dockerfile based on `lnl7/nix`, which is very similar: https://gist.github.com/MostAwesomeDude/be0bc41792d49a2887dd4100b54ca08a

Because of the way that Nix works, you're unlikely to be able to actually generate bleeding-edge images without annoying amounts of work inside Docker; the images that you generate will probably end up lagging somewhat behind the rest of nixpkgs-unstable. This isn't the end of the world, but it suggests that we should carry a Nix expression for Mailpile inside nixpkgs if folks would like to build it with e.g. a freshly-patched OpenSSL, and that we should build that expression from Hydra.

**Comment by yorickvP** [comment]
Nearly all of the packaging work is having a nix file for it, preferably in nixpkgs. Build-bots are optional, since nix can compile from source. Having a binary package is nice but unnecessary (especially since this is a python package), so most of what your build-bot would do is verify that it builds (which is likely, since it's a python package without any tests).

I guess you could provide a channel, but it will lag behind and be a pain for everyone to install and use.

You should probably just [accept](https://aur.archlinux.org/packages/mailpile/) [downstreams](https://gpo.zugaina.org/mail-client/Mailpile) [exist](https://pypi.python.org/pypi/mailpile/) and people want to use your software, and not stop them from doing that. You could add a version check that yells at them, if you have to. Or even take down old releases so people can't install them (which I hate). You are also welcome to pull-request version bumps yourself, they usually end up in the hands of the people within a few days after getting into the repo. We have some security people ( @grahamc) as well, for anything time-sensitive.

**Comment by BjarniRunar** [comment]
@yorickvP I've already explained why I don't want to encourage downstream packaging *at this point in time*. Telling me to, in effect, just suck it up and work harder, is absolutely not helpful.

Yes, downstream exists. And it's problematic, and I've asked you politely to please not be part of the problem. Basically, if there's no automated way to provide fully up-to-date packages to NixOS users, then please wait until one of our release candidates gets blessed as 1.0.

**Comment by bjornfor** [comment]
Perhaps try building distro-agnostic packages? For example, https://github.com/matthewbauer/nix-bundle?

**Comment by volth** [comment]
@BjarniRunar Or you may maintain a .nix-file (it is a small text file, not binary blob) in your repo, like many products maintain a ```Dockerfile``` in their repo (and some also ```.nix```: https://github.com/scala-native/scala-native/tree/0df05ab9ce20336fb65d8814a403cf2d6c7ef393/bin)

**Comment by yorickvP** [comment]
I updated the [mailpile package for local use](https://github.com/yorickvP/dotfiles/blob/master/nix/.config/nixpkgs/pkgs/mailpile.nix), so I could at least check out what mailpile is about. It'll stop working once another 1.0 release comes out.

**Comment by BjarniRunar** [comment]
@bjornfor , @volth - those both sound like interesting options. I won't have time to work on them myself in the near future, but I would accept contributions. That's especially applicable to a .nix file I guess. How does that work? Do people just "nix install https://github.com/..../foo/.nix" ?  How are updates detected/handled?

**Comment by bjornfor** [comment]
> those both sound like interesting options

I think if you don't like the idea of downstreams, and especially _multiple_ downstreams, I think you should strongly consider nix-bundle. That should allow you to build one bundle and distribute to all modern Linux' systems. You'd still have to write a .nix file to describe how to build mailpile, but that effort benefits all your users. (Even yourself!)

> Do people just "nix install https://github.com/..../foo/.nix" ? How are updates detected/handled?

No, it's not that simple, unfortunately. End users would then have to write a small Nix expression that imports the expression from you (https://github.com/foo/default.nix"). It is up to the user whether to fetch the latest file every time they update their system, or to lock down the initial version they built by specifying a sha256 sum for the upstream .nix file.

**Comment by yorickvP** [comment]
@bjornfor well, you could make a tarball and include it. For example: `nix-env -f https://pub.yori.cc/mailpile.tar -i` will install mailpile.

**Comment by bjornfor** [comment]
Good point!

One problem:
```
$ nix-env -f https://pub.yori.cc/mailpile.tar -i
downloading ‘https://pub.yori.cc/mailpile.tar’... [0/0 KiB, 0.0 KiB/s]
unpacking ‘https://pub.yori.cc/mailpile.tar’...
installing ‘mailpile-1.0rc1’
error: invalid character ‘/’ in name ‘Mailpile-release/1.0-src’
(use ‘--show-trace’ to show detailed location information)
```

**Comment by yorickvP** [comment]
Works for me. Maybe it's a platform issue?

**Comment by bjornfor** [comment]
IDK. I'm running Nix 1.11.13 on NixOS 17.03.

**Comment by yorickvP** [comment]
Hm. maybe it doesn't like the / in the rev. (that'd be a bug, should investigate). I updated it to use the commit hash. It should work now.

**Comment by bjornfor** [comment]
Yes, it works! (I had to manually download the tarball and say "nix-env -f file://$PWD/mailpile.tar -i" because it seems nix cached the previous tarball somewhere.)

**Comment by Mic92** [comment]
@makefu uses @buildbot to update a nix channel. This could be a way of keeping the package up-to-date.
